// capnp-ts/lib/index.js
"use strict";
/**
 * @author jdiaz5513
 */
Object.defineProperty(exports, "__esModule", { value: true });
var serialization_1 = require("./serialization");
exports.ListElementSize = serialization_1.ListElementSize;
exports.Message = serialization_1.Message;
exports.ObjectSize = serialization_1.ObjectSize;
exports.readRawPointer = serialization_1.readRawPointer;
exports.AnyPointerList = serialization_1.AnyPointerList;
exports.BoolList = serialization_1.BoolList;
exports.CompositeList = serialization_1.CompositeList;
exports.Data = serialization_1.Data;
exports.DataList = serialization_1.DataList;
exports.Float32List = serialization_1.Float32List;
exports.Float64List = serialization_1.Float64List;
exports.Int16List = serialization_1.Int16List;
exports.Int32List = serialization_1.Int32List;
exports.Int64List = serialization_1.Int64List;
exports.Int8List = serialization_1.Int8List;
exports.Interface = serialization_1.Interface;
exports.InterfaceList = serialization_1.InterfaceList;
exports.List = serialization_1.List;
exports.Orphan = serialization_1.Orphan;
exports.PointerList = serialization_1.PointerList;
exports.PointerType = serialization_1.PointerType;
exports.Pointer = serialization_1.Pointer;
exports.Struct = serialization_1.Struct;
exports.Text = serialization_1.Text;
exports.TextList = serialization_1.TextList;
exports.Uint16List = serialization_1.Uint16List;
exports.Uint32List = serialization_1.Uint32List;
exports.Uint64List = serialization_1.Uint64List;
exports.Uint8List = serialization_1.Uint8List;
exports.VoidList = serialization_1.VoidList;
exports.Void = serialization_1.Void;
exports.getBitMask = serialization_1.getBitMask;
exports.getFloat32Mask = serialization_1.getFloat32Mask;
exports.getFloat64Mask = serialization_1.getFloat64Mask;
exports.getInt16Mask = serialization_1.getInt16Mask;
exports.getInt32Mask = serialization_1.getInt32Mask;
exports.getInt64Mask = serialization_1.getInt64Mask;
exports.getInt8Mask = serialization_1.getInt8Mask;
exports.getUint16Mask = serialization_1.getUint16Mask;
exports.getUint32Mask = serialization_1.getUint32Mask;
exports.getUint64Mask = serialization_1.getUint64Mask;
exports.getUint8Mask = serialization_1.getUint8Mask;
var types_1 = require("./types");
exports.Int64 = types_1.Int64;
exports.Uint64 = types_1.Uint64;
