// capnp-ts/lib/std/persistent.capnp.js
"use strict";
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var index_1 = require("../index");
exports._capnpFileId = "b8630836983feed7";
var Persistent_SaveParams = /** @class */ (function (_super) {
    tslib_1.__extends(Persistent_SaveParams, _super);
    function Persistent_SaveParams() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Persistent_SaveParams.prototype.adoptSealFor = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Persistent_SaveParams.prototype.disownSealFor = function () { return index_1.Struct.disown(this.getSealFor()); };
    Persistent_SaveParams.prototype.getSealFor = function () { return index_1.Struct.getPointer(0, this); };
    Persistent_SaveParams.prototype.hasSealFor = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Persistent_SaveParams.prototype.setSealFor = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Persistent_SaveParams.prototype.toString = function () { return "Persistent_SaveParams_" + _super.prototype.toString.call(this); };
    Persistent_SaveParams._capnp = { displayName: "SaveParams", id: "f76fba59183073a5", size: new index_1.ObjectSize(0, 1) };
    return Persistent_SaveParams;
}(index_1.Struct));
exports.Persistent_SaveParams = Persistent_SaveParams;
var Persistent_SaveResults = /** @class */ (function (_super) {
    tslib_1.__extends(Persistent_SaveResults, _super);
    function Persistent_SaveResults() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Persistent_SaveResults.prototype.adoptSturdyRef = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Persistent_SaveResults.prototype.disownSturdyRef = function () { return index_1.Struct.disown(this.getSturdyRef()); };
    Persistent_SaveResults.prototype.getSturdyRef = function () { return index_1.Struct.getPointer(0, this); };
    Persistent_SaveResults.prototype.hasSturdyRef = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Persistent_SaveResults.prototype.setSturdyRef = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Persistent_SaveResults.prototype.toString = function () { return "Persistent_SaveResults_" + _super.prototype.toString.call(this); };
    Persistent_SaveResults._capnp = { displayName: "SaveResults", id: "b76848c18c40efbf", size: new index_1.ObjectSize(0, 1) };
    return Persistent_SaveResults;
}(index_1.Struct));
exports.Persistent_SaveResults = Persistent_SaveResults;
var Persistent = /** @class */ (function (_super) {
    tslib_1.__extends(Persistent, _super);
    function Persistent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Persistent.prototype.toString = function () { return "Persistent_" + _super.prototype.toString.call(this); };
    Persistent.SaveParams = Persistent_SaveParams;
    Persistent.SaveResults = Persistent_SaveResults;
    Persistent._capnp = { displayName: "Persistent", id: "c8cb212fcd9f5691", size: new index_1.ObjectSize(0, 0) };
    return Persistent;
}(index_1.Struct));
exports.Persistent = Persistent;
var RealmGateway = /** @class */ (function (_super) {
    tslib_1.__extends(RealmGateway, _super);
    function RealmGateway() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RealmGateway.prototype.toString = function () { return "RealmGateway_" + _super.prototype.toString.call(this); };
    RealmGateway._capnp = { displayName: "RealmGateway", id: "84ff286cd00a3ed4", size: new index_1.ObjectSize(0, 0) };
    return RealmGateway;
}(index_1.Struct));
exports.RealmGateway = RealmGateway;
