// capnp-ts/lib/std/rpc-twoparty.capnp.js
"use strict";
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var index_1 = require("../index");
exports._capnpFileId = "a184c7885cdaf2a1";
var Side;
(function (Side) {
    Side[Side["SERVER"] = 0] = "SERVER";
    Side[Side["CLIENT"] = 1] = "CLIENT";
})(Side = exports.Side || (exports.Side = {}));
var VatId = /** @class */ (function (_super) {
    tslib_1.__extends(VatId, _super);
    function VatId() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    VatId.prototype.getSide = function () { return index_1.Struct.getUint16(0, this); };
    VatId.prototype.setSide = function (value) { index_1.Struct.setUint16(0, value, this); };
    VatId.prototype.toString = function () { return "VatId_" + _super.prototype.toString.call(this); };
    VatId._capnp = { displayName: "VatId", id: "d20b909fee733a8e", size: new index_1.ObjectSize(8, 0) };
    return VatId;
}(index_1.Struct));
exports.VatId = VatId;
var ProvisionId = /** @class */ (function (_super) {
    tslib_1.__extends(ProvisionId, _super);
    function ProvisionId() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ProvisionId.prototype.getJoinId = function () { return index_1.Struct.getUint32(0, this); };
    ProvisionId.prototype.setJoinId = function (value) { index_1.Struct.setUint32(0, value, this); };
    ProvisionId.prototype.toString = function () { return "ProvisionId_" + _super.prototype.toString.call(this); };
    ProvisionId._capnp = { displayName: "ProvisionId", id: "b88d09a9c5f39817", size: new index_1.ObjectSize(8, 0) };
    return ProvisionId;
}(index_1.Struct));
exports.ProvisionId = ProvisionId;
var RecipientId = /** @class */ (function (_super) {
    tslib_1.__extends(RecipientId, _super);
    function RecipientId() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RecipientId.prototype.toString = function () { return "RecipientId_" + _super.prototype.toString.call(this); };
    RecipientId._capnp = { displayName: "RecipientId", id: "89f389b6fd4082c1", size: new index_1.ObjectSize(0, 0) };
    return RecipientId;
}(index_1.Struct));
exports.RecipientId = RecipientId;
var ThirdPartyCapId = /** @class */ (function (_super) {
    tslib_1.__extends(ThirdPartyCapId, _super);
    function ThirdPartyCapId() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ThirdPartyCapId.prototype.toString = function () { return "ThirdPartyCapId_" + _super.prototype.toString.call(this); };
    ThirdPartyCapId._capnp = { displayName: "ThirdPartyCapId", id: "b47f4979672cb59d", size: new index_1.ObjectSize(0, 0) };
    return ThirdPartyCapId;
}(index_1.Struct));
exports.ThirdPartyCapId = ThirdPartyCapId;
var JoinKeyPart = /** @class */ (function (_super) {
    tslib_1.__extends(JoinKeyPart, _super);
    function JoinKeyPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JoinKeyPart.prototype.getJoinId = function () { return index_1.Struct.getUint32(0, this); };
    JoinKeyPart.prototype.setJoinId = function (value) { index_1.Struct.setUint32(0, value, this); };
    JoinKeyPart.prototype.getPartCount = function () { return index_1.Struct.getUint16(4, this); };
    JoinKeyPart.prototype.setPartCount = function (value) { index_1.Struct.setUint16(4, value, this); };
    JoinKeyPart.prototype.getPartNum = function () { return index_1.Struct.getUint16(6, this); };
    JoinKeyPart.prototype.setPartNum = function (value) { index_1.Struct.setUint16(6, value, this); };
    JoinKeyPart.prototype.toString = function () { return "JoinKeyPart_" + _super.prototype.toString.call(this); };
    JoinKeyPart._capnp = { displayName: "JoinKeyPart", id: "95b29059097fca83", size: new index_1.ObjectSize(8, 0) };
    return JoinKeyPart;
}(index_1.Struct));
exports.JoinKeyPart = JoinKeyPart;
var JoinResult = /** @class */ (function (_super) {
    tslib_1.__extends(JoinResult, _super);
    function JoinResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JoinResult.prototype.getJoinId = function () { return index_1.Struct.getUint32(0, this); };
    JoinResult.prototype.setJoinId = function (value) { index_1.Struct.setUint32(0, value, this); };
    JoinResult.prototype.getSucceeded = function () { return index_1.Struct.getBit(32, this); };
    JoinResult.prototype.setSucceeded = function (value) { index_1.Struct.setBit(32, value, this); };
    JoinResult.prototype.adoptCap = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    JoinResult.prototype.disownCap = function () { return index_1.Struct.disown(this.getCap()); };
    JoinResult.prototype.getCap = function () { return index_1.Struct.getPointer(0, this); };
    JoinResult.prototype.hasCap = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    JoinResult.prototype.setCap = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    JoinResult.prototype.toString = function () { return "JoinResult_" + _super.prototype.toString.call(this); };
    JoinResult._capnp = { displayName: "JoinResult", id: "9d263a3630b7ebee", size: new index_1.ObjectSize(8, 1) };
    return JoinResult;
}(index_1.Struct));
exports.JoinResult = JoinResult;
