// capnp-ts/lib/std/rpc.capnp.js
"use strict";
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * This file has been automatically generated by the [capnpc-ts utility](https://github.com/jdiaz5513/capnp-ts).
 */
var capnp = require("../index");
var index_1 = require("../index");
exports._capnpFileId = "b312981b2552a250";
var Message_Which;
(function (Message_Which) {
    Message_Which[Message_Which["UNIMPLEMENTED"] = 0] = "UNIMPLEMENTED";
    Message_Which[Message_Which["ABORT"] = 1] = "ABORT";
    Message_Which[Message_Which["BOOTSTRAP"] = 8] = "BOOTSTRAP";
    Message_Which[Message_Which["CALL"] = 2] = "CALL";
    Message_Which[Message_Which["RETURN"] = 3] = "RETURN";
    Message_Which[Message_Which["FINISH"] = 4] = "FINISH";
    Message_Which[Message_Which["RESOLVE"] = 5] = "RESOLVE";
    Message_Which[Message_Which["RELEASE"] = 6] = "RELEASE";
    Message_Which[Message_Which["DISEMBARGO"] = 13] = "DISEMBARGO";
    Message_Which[Message_Which["OBSOLETE_SAVE"] = 7] = "OBSOLETE_SAVE";
    Message_Which[Message_Which["OBSOLETE_DELETE"] = 9] = "OBSOLETE_DELETE";
    Message_Which[Message_Which["PROVIDE"] = 10] = "PROVIDE";
    Message_Which[Message_Which["ACCEPT"] = 11] = "ACCEPT";
    Message_Which[Message_Which["JOIN"] = 12] = "JOIN";
})(Message_Which = exports.Message_Which || (exports.Message_Which = {}));
var Message = /** @class */ (function (_super) {
    tslib_1.__extends(Message, _super);
    function Message() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Message.prototype.adoptUnimplemented = function (value) {
        index_1.Struct.setUint16(0, 0, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownUnimplemented = function () { return index_1.Struct.disown(this.getUnimplemented()); };
    Message.prototype.getUnimplemented = function () {
        index_1.Struct.testWhich("unimplemented", index_1.Struct.getUint16(0, this), 0, this);
        return index_1.Struct.getStruct(0, Message, this);
    };
    Message.prototype.hasUnimplemented = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initUnimplemented = function () {
        index_1.Struct.setUint16(0, 0, this);
        return index_1.Struct.initStructAt(0, Message, this);
    };
    Message.prototype.isUnimplemented = function () { return index_1.Struct.getUint16(0, this) === 0; };
    Message.prototype.setUnimplemented = function (value) {
        index_1.Struct.setUint16(0, 0, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptAbort = function (value) {
        index_1.Struct.setUint16(0, 1, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownAbort = function () { return index_1.Struct.disown(this.getAbort()); };
    Message.prototype.getAbort = function () {
        index_1.Struct.testWhich("abort", index_1.Struct.getUint16(0, this), 1, this);
        return index_1.Struct.getStruct(0, Exception, this);
    };
    Message.prototype.hasAbort = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initAbort = function () {
        index_1.Struct.setUint16(0, 1, this);
        return index_1.Struct.initStructAt(0, Exception, this);
    };
    Message.prototype.isAbort = function () { return index_1.Struct.getUint16(0, this) === 1; };
    Message.prototype.setAbort = function (value) {
        index_1.Struct.setUint16(0, 1, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptBootstrap = function (value) {
        index_1.Struct.setUint16(0, 8, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownBootstrap = function () { return index_1.Struct.disown(this.getBootstrap()); };
    Message.prototype.getBootstrap = function () {
        index_1.Struct.testWhich("bootstrap", index_1.Struct.getUint16(0, this), 8, this);
        return index_1.Struct.getStruct(0, Bootstrap, this);
    };
    Message.prototype.hasBootstrap = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initBootstrap = function () {
        index_1.Struct.setUint16(0, 8, this);
        return index_1.Struct.initStructAt(0, Bootstrap, this);
    };
    Message.prototype.isBootstrap = function () { return index_1.Struct.getUint16(0, this) === 8; };
    Message.prototype.setBootstrap = function (value) {
        index_1.Struct.setUint16(0, 8, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptCall = function (value) {
        index_1.Struct.setUint16(0, 2, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownCall = function () { return index_1.Struct.disown(this.getCall()); };
    Message.prototype.getCall = function () {
        index_1.Struct.testWhich("call", index_1.Struct.getUint16(0, this), 2, this);
        return index_1.Struct.getStruct(0, Call, this);
    };
    Message.prototype.hasCall = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initCall = function () {
        index_1.Struct.setUint16(0, 2, this);
        return index_1.Struct.initStructAt(0, Call, this);
    };
    Message.prototype.isCall = function () { return index_1.Struct.getUint16(0, this) === 2; };
    Message.prototype.setCall = function (value) {
        index_1.Struct.setUint16(0, 2, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptReturn = function (value) {
        index_1.Struct.setUint16(0, 3, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownReturn = function () { return index_1.Struct.disown(this.getReturn()); };
    Message.prototype.getReturn = function () {
        index_1.Struct.testWhich("return", index_1.Struct.getUint16(0, this), 3, this);
        return index_1.Struct.getStruct(0, Return, this);
    };
    Message.prototype.hasReturn = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initReturn = function () {
        index_1.Struct.setUint16(0, 3, this);
        return index_1.Struct.initStructAt(0, Return, this);
    };
    Message.prototype.isReturn = function () { return index_1.Struct.getUint16(0, this) === 3; };
    Message.prototype.setReturn = function (value) {
        index_1.Struct.setUint16(0, 3, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptFinish = function (value) {
        index_1.Struct.setUint16(0, 4, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownFinish = function () { return index_1.Struct.disown(this.getFinish()); };
    Message.prototype.getFinish = function () {
        index_1.Struct.testWhich("finish", index_1.Struct.getUint16(0, this), 4, this);
        return index_1.Struct.getStruct(0, Finish, this);
    };
    Message.prototype.hasFinish = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initFinish = function () {
        index_1.Struct.setUint16(0, 4, this);
        return index_1.Struct.initStructAt(0, Finish, this);
    };
    Message.prototype.isFinish = function () { return index_1.Struct.getUint16(0, this) === 4; };
    Message.prototype.setFinish = function (value) {
        index_1.Struct.setUint16(0, 4, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptResolve = function (value) {
        index_1.Struct.setUint16(0, 5, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownResolve = function () { return index_1.Struct.disown(this.getResolve()); };
    Message.prototype.getResolve = function () {
        index_1.Struct.testWhich("resolve", index_1.Struct.getUint16(0, this), 5, this);
        return index_1.Struct.getStruct(0, Resolve, this);
    };
    Message.prototype.hasResolve = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initResolve = function () {
        index_1.Struct.setUint16(0, 5, this);
        return index_1.Struct.initStructAt(0, Resolve, this);
    };
    Message.prototype.isResolve = function () { return index_1.Struct.getUint16(0, this) === 5; };
    Message.prototype.setResolve = function (value) {
        index_1.Struct.setUint16(0, 5, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptRelease = function (value) {
        index_1.Struct.setUint16(0, 6, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownRelease = function () { return index_1.Struct.disown(this.getRelease()); };
    Message.prototype.getRelease = function () {
        index_1.Struct.testWhich("release", index_1.Struct.getUint16(0, this), 6, this);
        return index_1.Struct.getStruct(0, Release, this);
    };
    Message.prototype.hasRelease = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initRelease = function () {
        index_1.Struct.setUint16(0, 6, this);
        return index_1.Struct.initStructAt(0, Release, this);
    };
    Message.prototype.isRelease = function () { return index_1.Struct.getUint16(0, this) === 6; };
    Message.prototype.setRelease = function (value) {
        index_1.Struct.setUint16(0, 6, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptDisembargo = function (value) {
        index_1.Struct.setUint16(0, 13, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownDisembargo = function () { return index_1.Struct.disown(this.getDisembargo()); };
    Message.prototype.getDisembargo = function () {
        index_1.Struct.testWhich("disembargo", index_1.Struct.getUint16(0, this), 13, this);
        return index_1.Struct.getStruct(0, Disembargo, this);
    };
    Message.prototype.hasDisembargo = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initDisembargo = function () {
        index_1.Struct.setUint16(0, 13, this);
        return index_1.Struct.initStructAt(0, Disembargo, this);
    };
    Message.prototype.isDisembargo = function () { return index_1.Struct.getUint16(0, this) === 13; };
    Message.prototype.setDisembargo = function (value) {
        index_1.Struct.setUint16(0, 13, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptObsoleteSave = function (value) {
        index_1.Struct.setUint16(0, 7, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownObsoleteSave = function () { return index_1.Struct.disown(this.getObsoleteSave()); };
    Message.prototype.getObsoleteSave = function () {
        index_1.Struct.testWhich("obsoleteSave", index_1.Struct.getUint16(0, this), 7, this);
        return index_1.Struct.getPointer(0, this);
    };
    Message.prototype.hasObsoleteSave = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.isObsoleteSave = function () { return index_1.Struct.getUint16(0, this) === 7; };
    Message.prototype.setObsoleteSave = function (value) {
        index_1.Struct.setUint16(0, 7, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptObsoleteDelete = function (value) {
        index_1.Struct.setUint16(0, 9, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownObsoleteDelete = function () { return index_1.Struct.disown(this.getObsoleteDelete()); };
    Message.prototype.getObsoleteDelete = function () {
        index_1.Struct.testWhich("obsoleteDelete", index_1.Struct.getUint16(0, this), 9, this);
        return index_1.Struct.getPointer(0, this);
    };
    Message.prototype.hasObsoleteDelete = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.isObsoleteDelete = function () { return index_1.Struct.getUint16(0, this) === 9; };
    Message.prototype.setObsoleteDelete = function (value) {
        index_1.Struct.setUint16(0, 9, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptProvide = function (value) {
        index_1.Struct.setUint16(0, 10, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownProvide = function () { return index_1.Struct.disown(this.getProvide()); };
    Message.prototype.getProvide = function () {
        index_1.Struct.testWhich("provide", index_1.Struct.getUint16(0, this), 10, this);
        return index_1.Struct.getStruct(0, Provide, this);
    };
    Message.prototype.hasProvide = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initProvide = function () {
        index_1.Struct.setUint16(0, 10, this);
        return index_1.Struct.initStructAt(0, Provide, this);
    };
    Message.prototype.isProvide = function () { return index_1.Struct.getUint16(0, this) === 10; };
    Message.prototype.setProvide = function (value) {
        index_1.Struct.setUint16(0, 10, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptAccept = function (value) {
        index_1.Struct.setUint16(0, 11, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownAccept = function () { return index_1.Struct.disown(this.getAccept()); };
    Message.prototype.getAccept = function () {
        index_1.Struct.testWhich("accept", index_1.Struct.getUint16(0, this), 11, this);
        return index_1.Struct.getStruct(0, Accept, this);
    };
    Message.prototype.hasAccept = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initAccept = function () {
        index_1.Struct.setUint16(0, 11, this);
        return index_1.Struct.initStructAt(0, Accept, this);
    };
    Message.prototype.isAccept = function () { return index_1.Struct.getUint16(0, this) === 11; };
    Message.prototype.setAccept = function (value) {
        index_1.Struct.setUint16(0, 11, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.adoptJoin = function (value) {
        index_1.Struct.setUint16(0, 12, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.disownJoin = function () { return index_1.Struct.disown(this.getJoin()); };
    Message.prototype.getJoin = function () {
        index_1.Struct.testWhich("join", index_1.Struct.getUint16(0, this), 12, this);
        return index_1.Struct.getStruct(0, Join, this);
    };
    Message.prototype.hasJoin = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Message.prototype.initJoin = function () {
        index_1.Struct.setUint16(0, 12, this);
        return index_1.Struct.initStructAt(0, Join, this);
    };
    Message.prototype.isJoin = function () { return index_1.Struct.getUint16(0, this) === 12; };
    Message.prototype.setJoin = function (value) {
        index_1.Struct.setUint16(0, 12, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Message.prototype.toString = function () { return "Message_" + _super.prototype.toString.call(this); };
    Message.prototype.which = function () { return index_1.Struct.getUint16(0, this); };
    Message.UNIMPLEMENTED = Message_Which.UNIMPLEMENTED;
    Message.ABORT = Message_Which.ABORT;
    Message.BOOTSTRAP = Message_Which.BOOTSTRAP;
    Message.CALL = Message_Which.CALL;
    Message.RETURN = Message_Which.RETURN;
    Message.FINISH = Message_Which.FINISH;
    Message.RESOLVE = Message_Which.RESOLVE;
    Message.RELEASE = Message_Which.RELEASE;
    Message.DISEMBARGO = Message_Which.DISEMBARGO;
    Message.OBSOLETE_SAVE = Message_Which.OBSOLETE_SAVE;
    Message.OBSOLETE_DELETE = Message_Which.OBSOLETE_DELETE;
    Message.PROVIDE = Message_Which.PROVIDE;
    Message.ACCEPT = Message_Which.ACCEPT;
    Message.JOIN = Message_Which.JOIN;
    Message._capnp = { displayName: "Message", id: "91b79f1f808db032", size: new index_1.ObjectSize(8, 1) };
    return Message;
}(index_1.Struct));
exports.Message = Message;
var Bootstrap = /** @class */ (function (_super) {
    tslib_1.__extends(Bootstrap, _super);
    function Bootstrap() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Bootstrap.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    Bootstrap.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Bootstrap.prototype.adoptDeprecatedObjectId = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Bootstrap.prototype.disownDeprecatedObjectId = function () { return index_1.Struct.disown(this.getDeprecatedObjectId()); };
    Bootstrap.prototype.getDeprecatedObjectId = function () { return index_1.Struct.getPointer(0, this); };
    Bootstrap.prototype.hasDeprecatedObjectId = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Bootstrap.prototype.setDeprecatedObjectId = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Bootstrap.prototype.toString = function () { return "Bootstrap_" + _super.prototype.toString.call(this); };
    Bootstrap._capnp = { displayName: "Bootstrap", id: "e94ccf8031176ec4", size: new index_1.ObjectSize(8, 1) };
    return Bootstrap;
}(index_1.Struct));
exports.Bootstrap = Bootstrap;
var Call_SendResultsTo_Which;
(function (Call_SendResultsTo_Which) {
    Call_SendResultsTo_Which[Call_SendResultsTo_Which["CALLER"] = 0] = "CALLER";
    Call_SendResultsTo_Which[Call_SendResultsTo_Which["YOURSELF"] = 1] = "YOURSELF";
    Call_SendResultsTo_Which[Call_SendResultsTo_Which["THIRD_PARTY"] = 2] = "THIRD_PARTY";
})(Call_SendResultsTo_Which = exports.Call_SendResultsTo_Which || (exports.Call_SendResultsTo_Which = {}));
var Call_SendResultsTo = /** @class */ (function (_super) {
    tslib_1.__extends(Call_SendResultsTo, _super);
    function Call_SendResultsTo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Call_SendResultsTo.prototype.isCaller = function () { return index_1.Struct.getUint16(6, this) === 0; };
    Call_SendResultsTo.prototype.setCaller = function () { index_1.Struct.setUint16(6, 0, this); };
    Call_SendResultsTo.prototype.isYourself = function () { return index_1.Struct.getUint16(6, this) === 1; };
    Call_SendResultsTo.prototype.setYourself = function () { index_1.Struct.setUint16(6, 1, this); };
    Call_SendResultsTo.prototype.adoptThirdParty = function (value) {
        index_1.Struct.setUint16(6, 2, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(2, this));
    };
    Call_SendResultsTo.prototype.disownThirdParty = function () { return index_1.Struct.disown(this.getThirdParty()); };
    Call_SendResultsTo.prototype.getThirdParty = function () {
        index_1.Struct.testWhich("thirdParty", index_1.Struct.getUint16(6, this), 2, this);
        return index_1.Struct.getPointer(2, this);
    };
    Call_SendResultsTo.prototype.hasThirdParty = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(2, this)); };
    Call_SendResultsTo.prototype.isThirdParty = function () { return index_1.Struct.getUint16(6, this) === 2; };
    Call_SendResultsTo.prototype.setThirdParty = function (value) {
        index_1.Struct.setUint16(6, 2, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(2, this));
    };
    Call_SendResultsTo.prototype.toString = function () { return "Call_SendResultsTo_" + _super.prototype.toString.call(this); };
    Call_SendResultsTo.prototype.which = function () { return index_1.Struct.getUint16(6, this); };
    Call_SendResultsTo.CALLER = Call_SendResultsTo_Which.CALLER;
    Call_SendResultsTo.YOURSELF = Call_SendResultsTo_Which.YOURSELF;
    Call_SendResultsTo.THIRD_PARTY = Call_SendResultsTo_Which.THIRD_PARTY;
    Call_SendResultsTo._capnp = { displayName: "sendResultsTo", id: "dae8b0f61aab5f99", size: new index_1.ObjectSize(24, 3) };
    return Call_SendResultsTo;
}(index_1.Struct));
exports.Call_SendResultsTo = Call_SendResultsTo;
var Call = /** @class */ (function (_super) {
    tslib_1.__extends(Call, _super);
    function Call() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Call.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    Call.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Call.prototype.adoptTarget = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Call.prototype.disownTarget = function () { return index_1.Struct.disown(this.getTarget()); };
    Call.prototype.getTarget = function () { return index_1.Struct.getStruct(0, MessageTarget, this); };
    Call.prototype.hasTarget = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Call.prototype.initTarget = function () { return index_1.Struct.initStructAt(0, MessageTarget, this); };
    Call.prototype.setTarget = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Call.prototype.getInterfaceId = function () { return index_1.Struct.getUint64(8, this); };
    Call.prototype.setInterfaceId = function (value) { index_1.Struct.setUint64(8, value, this); };
    Call.prototype.getMethodId = function () { return index_1.Struct.getUint16(4, this); };
    Call.prototype.setMethodId = function (value) { index_1.Struct.setUint16(4, value, this); };
    Call.prototype.getAllowThirdPartyTailCall = function () { return index_1.Struct.getBit(128, this, Call._capnp.defaultAllowThirdPartyTailCall); };
    Call.prototype.setAllowThirdPartyTailCall = function (value) { index_1.Struct.setBit(128, value, this); };
    Call.prototype.adoptParams = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(1, this)); };
    Call.prototype.disownParams = function () { return index_1.Struct.disown(this.getParams()); };
    Call.prototype.getParams = function () { return index_1.Struct.getStruct(1, Payload, this); };
    Call.prototype.hasParams = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(1, this)); };
    Call.prototype.initParams = function () { return index_1.Struct.initStructAt(1, Payload, this); };
    Call.prototype.setParams = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(1, this)); };
    Call.prototype.getSendResultsTo = function () { return index_1.Struct.getAs(Call_SendResultsTo, this); };
    Call.prototype.initSendResultsTo = function () { return index_1.Struct.getAs(Call_SendResultsTo, this); };
    Call.prototype.toString = function () { return "Call_" + _super.prototype.toString.call(this); };
    Call._capnp = { displayName: "Call", id: "836a53ce789d4cd4", size: new index_1.ObjectSize(24, 3), defaultAllowThirdPartyTailCall: capnp.getBitMask(false, 0) };
    return Call;
}(index_1.Struct));
exports.Call = Call;
var Return_Which;
(function (Return_Which) {
    Return_Which[Return_Which["RESULTS"] = 0] = "RESULTS";
    Return_Which[Return_Which["EXCEPTION"] = 1] = "EXCEPTION";
    Return_Which[Return_Which["CANCELED"] = 2] = "CANCELED";
    Return_Which[Return_Which["RESULTS_SENT_ELSEWHERE"] = 3] = "RESULTS_SENT_ELSEWHERE";
    Return_Which[Return_Which["TAKE_FROM_OTHER_QUESTION"] = 4] = "TAKE_FROM_OTHER_QUESTION";
    Return_Which[Return_Which["ACCEPT_FROM_THIRD_PARTY"] = 5] = "ACCEPT_FROM_THIRD_PARTY";
})(Return_Which = exports.Return_Which || (exports.Return_Which = {}));
var Return = /** @class */ (function (_super) {
    tslib_1.__extends(Return, _super);
    function Return() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Return.prototype.getAnswerId = function () { return index_1.Struct.getUint32(0, this); };
    Return.prototype.setAnswerId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Return.prototype.getReleaseParamCaps = function () { return index_1.Struct.getBit(32, this, Return._capnp.defaultReleaseParamCaps); };
    Return.prototype.setReleaseParamCaps = function (value) { index_1.Struct.setBit(32, value, this); };
    Return.prototype.adoptResults = function (value) {
        index_1.Struct.setUint16(6, 0, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Return.prototype.disownResults = function () { return index_1.Struct.disown(this.getResults()); };
    Return.prototype.getResults = function () {
        index_1.Struct.testWhich("results", index_1.Struct.getUint16(6, this), 0, this);
        return index_1.Struct.getStruct(0, Payload, this);
    };
    Return.prototype.hasResults = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Return.prototype.initResults = function () {
        index_1.Struct.setUint16(6, 0, this);
        return index_1.Struct.initStructAt(0, Payload, this);
    };
    Return.prototype.isResults = function () { return index_1.Struct.getUint16(6, this) === 0; };
    Return.prototype.setResults = function (value) {
        index_1.Struct.setUint16(6, 0, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Return.prototype.adoptException = function (value) {
        index_1.Struct.setUint16(6, 1, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Return.prototype.disownException = function () { return index_1.Struct.disown(this.getException()); };
    Return.prototype.getException = function () {
        index_1.Struct.testWhich("exception", index_1.Struct.getUint16(6, this), 1, this);
        return index_1.Struct.getStruct(0, Exception, this);
    };
    Return.prototype.hasException = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Return.prototype.initException = function () {
        index_1.Struct.setUint16(6, 1, this);
        return index_1.Struct.initStructAt(0, Exception, this);
    };
    Return.prototype.isException = function () { return index_1.Struct.getUint16(6, this) === 1; };
    Return.prototype.setException = function (value) {
        index_1.Struct.setUint16(6, 1, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Return.prototype.isCanceled = function () { return index_1.Struct.getUint16(6, this) === 2; };
    Return.prototype.setCanceled = function () { index_1.Struct.setUint16(6, 2, this); };
    Return.prototype.isResultsSentElsewhere = function () { return index_1.Struct.getUint16(6, this) === 3; };
    Return.prototype.setResultsSentElsewhere = function () { index_1.Struct.setUint16(6, 3, this); };
    Return.prototype.getTakeFromOtherQuestion = function () {
        index_1.Struct.testWhich("takeFromOtherQuestion", index_1.Struct.getUint16(6, this), 4, this);
        return index_1.Struct.getUint32(8, this);
    };
    Return.prototype.isTakeFromOtherQuestion = function () { return index_1.Struct.getUint16(6, this) === 4; };
    Return.prototype.setTakeFromOtherQuestion = function (value) {
        index_1.Struct.setUint16(6, 4, this);
        index_1.Struct.setUint32(8, value, this);
    };
    Return.prototype.adoptAcceptFromThirdParty = function (value) {
        index_1.Struct.setUint16(6, 5, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Return.prototype.disownAcceptFromThirdParty = function () { return index_1.Struct.disown(this.getAcceptFromThirdParty()); };
    Return.prototype.getAcceptFromThirdParty = function () {
        index_1.Struct.testWhich("acceptFromThirdParty", index_1.Struct.getUint16(6, this), 5, this);
        return index_1.Struct.getPointer(0, this);
    };
    Return.prototype.hasAcceptFromThirdParty = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Return.prototype.isAcceptFromThirdParty = function () { return index_1.Struct.getUint16(6, this) === 5; };
    Return.prototype.setAcceptFromThirdParty = function (value) {
        index_1.Struct.setUint16(6, 5, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Return.prototype.toString = function () { return "Return_" + _super.prototype.toString.call(this); };
    Return.prototype.which = function () { return index_1.Struct.getUint16(6, this); };
    Return.RESULTS = Return_Which.RESULTS;
    Return.EXCEPTION = Return_Which.EXCEPTION;
    Return.CANCELED = Return_Which.CANCELED;
    Return.RESULTS_SENT_ELSEWHERE = Return_Which.RESULTS_SENT_ELSEWHERE;
    Return.TAKE_FROM_OTHER_QUESTION = Return_Which.TAKE_FROM_OTHER_QUESTION;
    Return.ACCEPT_FROM_THIRD_PARTY = Return_Which.ACCEPT_FROM_THIRD_PARTY;
    Return._capnp = { displayName: "Return", id: "9e19b28d3db3573a", size: new index_1.ObjectSize(16, 1), defaultReleaseParamCaps: capnp.getBitMask(true, 0) };
    return Return;
}(index_1.Struct));
exports.Return = Return;
var Finish = /** @class */ (function (_super) {
    tslib_1.__extends(Finish, _super);
    function Finish() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Finish.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    Finish.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Finish.prototype.getReleaseResultCaps = function () { return index_1.Struct.getBit(32, this, Finish._capnp.defaultReleaseResultCaps); };
    Finish.prototype.setReleaseResultCaps = function (value) { index_1.Struct.setBit(32, value, this); };
    Finish.prototype.toString = function () { return "Finish_" + _super.prototype.toString.call(this); };
    Finish._capnp = { displayName: "Finish", id: "d37d2eb2c2f80e63", size: new index_1.ObjectSize(8, 0), defaultReleaseResultCaps: capnp.getBitMask(true, 0) };
    return Finish;
}(index_1.Struct));
exports.Finish = Finish;
var Resolve_Which;
(function (Resolve_Which) {
    Resolve_Which[Resolve_Which["CAP"] = 0] = "CAP";
    Resolve_Which[Resolve_Which["EXCEPTION"] = 1] = "EXCEPTION";
})(Resolve_Which = exports.Resolve_Which || (exports.Resolve_Which = {}));
var Resolve = /** @class */ (function (_super) {
    tslib_1.__extends(Resolve, _super);
    function Resolve() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Resolve.prototype.getPromiseId = function () { return index_1.Struct.getUint32(0, this); };
    Resolve.prototype.setPromiseId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Resolve.prototype.adoptCap = function (value) {
        index_1.Struct.setUint16(4, 0, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Resolve.prototype.disownCap = function () { return index_1.Struct.disown(this.getCap()); };
    Resolve.prototype.getCap = function () {
        index_1.Struct.testWhich("cap", index_1.Struct.getUint16(4, this), 0, this);
        return index_1.Struct.getStruct(0, CapDescriptor, this);
    };
    Resolve.prototype.hasCap = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Resolve.prototype.initCap = function () {
        index_1.Struct.setUint16(4, 0, this);
        return index_1.Struct.initStructAt(0, CapDescriptor, this);
    };
    Resolve.prototype.isCap = function () { return index_1.Struct.getUint16(4, this) === 0; };
    Resolve.prototype.setCap = function (value) {
        index_1.Struct.setUint16(4, 0, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Resolve.prototype.adoptException = function (value) {
        index_1.Struct.setUint16(4, 1, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    Resolve.prototype.disownException = function () { return index_1.Struct.disown(this.getException()); };
    Resolve.prototype.getException = function () {
        index_1.Struct.testWhich("exception", index_1.Struct.getUint16(4, this), 1, this);
        return index_1.Struct.getStruct(0, Exception, this);
    };
    Resolve.prototype.hasException = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Resolve.prototype.initException = function () {
        index_1.Struct.setUint16(4, 1, this);
        return index_1.Struct.initStructAt(0, Exception, this);
    };
    Resolve.prototype.isException = function () { return index_1.Struct.getUint16(4, this) === 1; };
    Resolve.prototype.setException = function (value) {
        index_1.Struct.setUint16(4, 1, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    Resolve.prototype.toString = function () { return "Resolve_" + _super.prototype.toString.call(this); };
    Resolve.prototype.which = function () { return index_1.Struct.getUint16(4, this); };
    Resolve.CAP = Resolve_Which.CAP;
    Resolve.EXCEPTION = Resolve_Which.EXCEPTION;
    Resolve._capnp = { displayName: "Resolve", id: "bbc29655fa89086e", size: new index_1.ObjectSize(8, 1) };
    return Resolve;
}(index_1.Struct));
exports.Resolve = Resolve;
var Release = /** @class */ (function (_super) {
    tslib_1.__extends(Release, _super);
    function Release() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Release.prototype.getId = function () { return index_1.Struct.getUint32(0, this); };
    Release.prototype.setId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Release.prototype.getReferenceCount = function () { return index_1.Struct.getUint32(4, this); };
    Release.prototype.setReferenceCount = function (value) { index_1.Struct.setUint32(4, value, this); };
    Release.prototype.toString = function () { return "Release_" + _super.prototype.toString.call(this); };
    Release._capnp = { displayName: "Release", id: "ad1a6c0d7dd07497", size: new index_1.ObjectSize(8, 0) };
    return Release;
}(index_1.Struct));
exports.Release = Release;
var Disembargo_Context_Which;
(function (Disembargo_Context_Which) {
    Disembargo_Context_Which[Disembargo_Context_Which["SENDER_LOOPBACK"] = 0] = "SENDER_LOOPBACK";
    Disembargo_Context_Which[Disembargo_Context_Which["RECEIVER_LOOPBACK"] = 1] = "RECEIVER_LOOPBACK";
    Disembargo_Context_Which[Disembargo_Context_Which["ACCEPT"] = 2] = "ACCEPT";
    Disembargo_Context_Which[Disembargo_Context_Which["PROVIDE"] = 3] = "PROVIDE";
})(Disembargo_Context_Which = exports.Disembargo_Context_Which || (exports.Disembargo_Context_Which = {}));
var Disembargo_Context = /** @class */ (function (_super) {
    tslib_1.__extends(Disembargo_Context, _super);
    function Disembargo_Context() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Disembargo_Context.prototype.getSenderLoopback = function () {
        index_1.Struct.testWhich("senderLoopback", index_1.Struct.getUint16(4, this), 0, this);
        return index_1.Struct.getUint32(0, this);
    };
    Disembargo_Context.prototype.isSenderLoopback = function () { return index_1.Struct.getUint16(4, this) === 0; };
    Disembargo_Context.prototype.setSenderLoopback = function (value) {
        index_1.Struct.setUint16(4, 0, this);
        index_1.Struct.setUint32(0, value, this);
    };
    Disembargo_Context.prototype.getReceiverLoopback = function () {
        index_1.Struct.testWhich("receiverLoopback", index_1.Struct.getUint16(4, this), 1, this);
        return index_1.Struct.getUint32(0, this);
    };
    Disembargo_Context.prototype.isReceiverLoopback = function () { return index_1.Struct.getUint16(4, this) === 1; };
    Disembargo_Context.prototype.setReceiverLoopback = function (value) {
        index_1.Struct.setUint16(4, 1, this);
        index_1.Struct.setUint32(0, value, this);
    };
    Disembargo_Context.prototype.isAccept = function () { return index_1.Struct.getUint16(4, this) === 2; };
    Disembargo_Context.prototype.setAccept = function () { index_1.Struct.setUint16(4, 2, this); };
    Disembargo_Context.prototype.getProvide = function () {
        index_1.Struct.testWhich("provide", index_1.Struct.getUint16(4, this), 3, this);
        return index_1.Struct.getUint32(0, this);
    };
    Disembargo_Context.prototype.isProvide = function () { return index_1.Struct.getUint16(4, this) === 3; };
    Disembargo_Context.prototype.setProvide = function (value) {
        index_1.Struct.setUint16(4, 3, this);
        index_1.Struct.setUint32(0, value, this);
    };
    Disembargo_Context.prototype.toString = function () { return "Disembargo_Context_" + _super.prototype.toString.call(this); };
    Disembargo_Context.prototype.which = function () { return index_1.Struct.getUint16(4, this); };
    Disembargo_Context.SENDER_LOOPBACK = Disembargo_Context_Which.SENDER_LOOPBACK;
    Disembargo_Context.RECEIVER_LOOPBACK = Disembargo_Context_Which.RECEIVER_LOOPBACK;
    Disembargo_Context.ACCEPT = Disembargo_Context_Which.ACCEPT;
    Disembargo_Context.PROVIDE = Disembargo_Context_Which.PROVIDE;
    Disembargo_Context._capnp = { displayName: "context", id: "d562b4df655bdd4d", size: new index_1.ObjectSize(8, 1) };
    return Disembargo_Context;
}(index_1.Struct));
exports.Disembargo_Context = Disembargo_Context;
var Disembargo = /** @class */ (function (_super) {
    tslib_1.__extends(Disembargo, _super);
    function Disembargo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Disembargo.prototype.adoptTarget = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Disembargo.prototype.disownTarget = function () { return index_1.Struct.disown(this.getTarget()); };
    Disembargo.prototype.getTarget = function () { return index_1.Struct.getStruct(0, MessageTarget, this); };
    Disembargo.prototype.hasTarget = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Disembargo.prototype.initTarget = function () { return index_1.Struct.initStructAt(0, MessageTarget, this); };
    Disembargo.prototype.setTarget = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Disembargo.prototype.getContext = function () { return index_1.Struct.getAs(Disembargo_Context, this); };
    Disembargo.prototype.initContext = function () { return index_1.Struct.getAs(Disembargo_Context, this); };
    Disembargo.prototype.toString = function () { return "Disembargo_" + _super.prototype.toString.call(this); };
    Disembargo._capnp = { displayName: "Disembargo", id: "f964368b0fbd3711", size: new index_1.ObjectSize(8, 1) };
    return Disembargo;
}(index_1.Struct));
exports.Disembargo = Disembargo;
var Provide = /** @class */ (function (_super) {
    tslib_1.__extends(Provide, _super);
    function Provide() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Provide.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    Provide.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Provide.prototype.adoptTarget = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Provide.prototype.disownTarget = function () { return index_1.Struct.disown(this.getTarget()); };
    Provide.prototype.getTarget = function () { return index_1.Struct.getStruct(0, MessageTarget, this); };
    Provide.prototype.hasTarget = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Provide.prototype.initTarget = function () { return index_1.Struct.initStructAt(0, MessageTarget, this); };
    Provide.prototype.setTarget = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Provide.prototype.adoptRecipient = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(1, this)); };
    Provide.prototype.disownRecipient = function () { return index_1.Struct.disown(this.getRecipient()); };
    Provide.prototype.getRecipient = function () { return index_1.Struct.getPointer(1, this); };
    Provide.prototype.hasRecipient = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(1, this)); };
    Provide.prototype.setRecipient = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(1, this)); };
    Provide.prototype.toString = function () { return "Provide_" + _super.prototype.toString.call(this); };
    Provide._capnp = { displayName: "Provide", id: "9c6a046bfbc1ac5a", size: new index_1.ObjectSize(8, 2) };
    return Provide;
}(index_1.Struct));
exports.Provide = Provide;
var Accept = /** @class */ (function (_super) {
    tslib_1.__extends(Accept, _super);
    function Accept() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Accept.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    Accept.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Accept.prototype.adoptProvision = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Accept.prototype.disownProvision = function () { return index_1.Struct.disown(this.getProvision()); };
    Accept.prototype.getProvision = function () { return index_1.Struct.getPointer(0, this); };
    Accept.prototype.hasProvision = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Accept.prototype.setProvision = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Accept.prototype.getEmbargo = function () { return index_1.Struct.getBit(32, this); };
    Accept.prototype.setEmbargo = function (value) { index_1.Struct.setBit(32, value, this); };
    Accept.prototype.toString = function () { return "Accept_" + _super.prototype.toString.call(this); };
    Accept._capnp = { displayName: "Accept", id: "d4c9b56290554016", size: new index_1.ObjectSize(8, 1) };
    return Accept;
}(index_1.Struct));
exports.Accept = Accept;
var Join = /** @class */ (function (_super) {
    tslib_1.__extends(Join, _super);
    function Join() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Join.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    Join.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    Join.prototype.adoptTarget = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Join.prototype.disownTarget = function () { return index_1.Struct.disown(this.getTarget()); };
    Join.prototype.getTarget = function () { return index_1.Struct.getStruct(0, MessageTarget, this); };
    Join.prototype.hasTarget = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Join.prototype.initTarget = function () { return index_1.Struct.initStructAt(0, MessageTarget, this); };
    Join.prototype.setTarget = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Join.prototype.adoptKeyPart = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(1, this)); };
    Join.prototype.disownKeyPart = function () { return index_1.Struct.disown(this.getKeyPart()); };
    Join.prototype.getKeyPart = function () { return index_1.Struct.getPointer(1, this); };
    Join.prototype.hasKeyPart = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(1, this)); };
    Join.prototype.setKeyPart = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(1, this)); };
    Join.prototype.toString = function () { return "Join_" + _super.prototype.toString.call(this); };
    Join._capnp = { displayName: "Join", id: "fbe1980490e001af", size: new index_1.ObjectSize(8, 2) };
    return Join;
}(index_1.Struct));
exports.Join = Join;
var MessageTarget_Which;
(function (MessageTarget_Which) {
    MessageTarget_Which[MessageTarget_Which["IMPORTED_CAP"] = 0] = "IMPORTED_CAP";
    MessageTarget_Which[MessageTarget_Which["PROMISED_ANSWER"] = 1] = "PROMISED_ANSWER";
})(MessageTarget_Which = exports.MessageTarget_Which || (exports.MessageTarget_Which = {}));
var MessageTarget = /** @class */ (function (_super) {
    tslib_1.__extends(MessageTarget, _super);
    function MessageTarget() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MessageTarget.prototype.getImportedCap = function () {
        index_1.Struct.testWhich("importedCap", index_1.Struct.getUint16(4, this), 0, this);
        return index_1.Struct.getUint32(0, this);
    };
    MessageTarget.prototype.isImportedCap = function () { return index_1.Struct.getUint16(4, this) === 0; };
    MessageTarget.prototype.setImportedCap = function (value) {
        index_1.Struct.setUint16(4, 0, this);
        index_1.Struct.setUint32(0, value, this);
    };
    MessageTarget.prototype.adoptPromisedAnswer = function (value) {
        index_1.Struct.setUint16(4, 1, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    MessageTarget.prototype.disownPromisedAnswer = function () { return index_1.Struct.disown(this.getPromisedAnswer()); };
    MessageTarget.prototype.getPromisedAnswer = function () {
        index_1.Struct.testWhich("promisedAnswer", index_1.Struct.getUint16(4, this), 1, this);
        return index_1.Struct.getStruct(0, PromisedAnswer, this);
    };
    MessageTarget.prototype.hasPromisedAnswer = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    MessageTarget.prototype.initPromisedAnswer = function () {
        index_1.Struct.setUint16(4, 1, this);
        return index_1.Struct.initStructAt(0, PromisedAnswer, this);
    };
    MessageTarget.prototype.isPromisedAnswer = function () { return index_1.Struct.getUint16(4, this) === 1; };
    MessageTarget.prototype.setPromisedAnswer = function (value) {
        index_1.Struct.setUint16(4, 1, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    MessageTarget.prototype.toString = function () { return "MessageTarget_" + _super.prototype.toString.call(this); };
    MessageTarget.prototype.which = function () { return index_1.Struct.getUint16(4, this); };
    MessageTarget.IMPORTED_CAP = MessageTarget_Which.IMPORTED_CAP;
    MessageTarget.PROMISED_ANSWER = MessageTarget_Which.PROMISED_ANSWER;
    MessageTarget._capnp = { displayName: "MessageTarget", id: "95bc14545813fbc1", size: new index_1.ObjectSize(8, 1) };
    return MessageTarget;
}(index_1.Struct));
exports.MessageTarget = MessageTarget;
var Payload = /** @class */ (function (_super) {
    tslib_1.__extends(Payload, _super);
    function Payload() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Payload.prototype.adoptContent = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    Payload.prototype.disownContent = function () { return index_1.Struct.disown(this.getContent()); };
    Payload.prototype.getContent = function () { return index_1.Struct.getPointer(0, this); };
    Payload.prototype.hasContent = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    Payload.prototype.setContent = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    Payload.prototype.adoptCapTable = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(1, this)); };
    Payload.prototype.disownCapTable = function () { return index_1.Struct.disown(this.getCapTable()); };
    Payload.prototype.getCapTable = function () { return index_1.Struct.getList(1, Payload._CapTable, this); };
    Payload.prototype.hasCapTable = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(1, this)); };
    Payload.prototype.initCapTable = function (length) { return index_1.Struct.initList(1, Payload._CapTable, length, this); };
    Payload.prototype.setCapTable = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(1, this)); };
    Payload.prototype.toString = function () { return "Payload_" + _super.prototype.toString.call(this); };
    Payload._capnp = { displayName: "Payload", id: "9a0e61223d96743b", size: new index_1.ObjectSize(0, 2) };
    return Payload;
}(index_1.Struct));
exports.Payload = Payload;
var CapDescriptor_Which;
(function (CapDescriptor_Which) {
    CapDescriptor_Which[CapDescriptor_Which["NONE"] = 0] = "NONE";
    CapDescriptor_Which[CapDescriptor_Which["SENDER_HOSTED"] = 1] = "SENDER_HOSTED";
    CapDescriptor_Which[CapDescriptor_Which["SENDER_PROMISE"] = 2] = "SENDER_PROMISE";
    CapDescriptor_Which[CapDescriptor_Which["RECEIVER_HOSTED"] = 3] = "RECEIVER_HOSTED";
    CapDescriptor_Which[CapDescriptor_Which["RECEIVER_ANSWER"] = 4] = "RECEIVER_ANSWER";
    CapDescriptor_Which[CapDescriptor_Which["THIRD_PARTY_HOSTED"] = 5] = "THIRD_PARTY_HOSTED";
})(CapDescriptor_Which = exports.CapDescriptor_Which || (exports.CapDescriptor_Which = {}));
var CapDescriptor = /** @class */ (function (_super) {
    tslib_1.__extends(CapDescriptor, _super);
    function CapDescriptor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CapDescriptor.prototype.isNone = function () { return index_1.Struct.getUint16(0, this) === 0; };
    CapDescriptor.prototype.setNone = function () { index_1.Struct.setUint16(0, 0, this); };
    CapDescriptor.prototype.getSenderHosted = function () {
        index_1.Struct.testWhich("senderHosted", index_1.Struct.getUint16(0, this), 1, this);
        return index_1.Struct.getUint32(4, this);
    };
    CapDescriptor.prototype.isSenderHosted = function () { return index_1.Struct.getUint16(0, this) === 1; };
    CapDescriptor.prototype.setSenderHosted = function (value) {
        index_1.Struct.setUint16(0, 1, this);
        index_1.Struct.setUint32(4, value, this);
    };
    CapDescriptor.prototype.getSenderPromise = function () {
        index_1.Struct.testWhich("senderPromise", index_1.Struct.getUint16(0, this), 2, this);
        return index_1.Struct.getUint32(4, this);
    };
    CapDescriptor.prototype.isSenderPromise = function () { return index_1.Struct.getUint16(0, this) === 2; };
    CapDescriptor.prototype.setSenderPromise = function (value) {
        index_1.Struct.setUint16(0, 2, this);
        index_1.Struct.setUint32(4, value, this);
    };
    CapDescriptor.prototype.getReceiverHosted = function () {
        index_1.Struct.testWhich("receiverHosted", index_1.Struct.getUint16(0, this), 3, this);
        return index_1.Struct.getUint32(4, this);
    };
    CapDescriptor.prototype.isReceiverHosted = function () { return index_1.Struct.getUint16(0, this) === 3; };
    CapDescriptor.prototype.setReceiverHosted = function (value) {
        index_1.Struct.setUint16(0, 3, this);
        index_1.Struct.setUint32(4, value, this);
    };
    CapDescriptor.prototype.adoptReceiverAnswer = function (value) {
        index_1.Struct.setUint16(0, 4, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    CapDescriptor.prototype.disownReceiverAnswer = function () { return index_1.Struct.disown(this.getReceiverAnswer()); };
    CapDescriptor.prototype.getReceiverAnswer = function () {
        index_1.Struct.testWhich("receiverAnswer", index_1.Struct.getUint16(0, this), 4, this);
        return index_1.Struct.getStruct(0, PromisedAnswer, this);
    };
    CapDescriptor.prototype.hasReceiverAnswer = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    CapDescriptor.prototype.initReceiverAnswer = function () {
        index_1.Struct.setUint16(0, 4, this);
        return index_1.Struct.initStructAt(0, PromisedAnswer, this);
    };
    CapDescriptor.prototype.isReceiverAnswer = function () { return index_1.Struct.getUint16(0, this) === 4; };
    CapDescriptor.prototype.setReceiverAnswer = function (value) {
        index_1.Struct.setUint16(0, 4, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    CapDescriptor.prototype.adoptThirdPartyHosted = function (value) {
        index_1.Struct.setUint16(0, 5, this);
        index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this));
    };
    CapDescriptor.prototype.disownThirdPartyHosted = function () { return index_1.Struct.disown(this.getThirdPartyHosted()); };
    CapDescriptor.prototype.getThirdPartyHosted = function () {
        index_1.Struct.testWhich("thirdPartyHosted", index_1.Struct.getUint16(0, this), 5, this);
        return index_1.Struct.getStruct(0, ThirdPartyCapDescriptor, this);
    };
    CapDescriptor.prototype.hasThirdPartyHosted = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    CapDescriptor.prototype.initThirdPartyHosted = function () {
        index_1.Struct.setUint16(0, 5, this);
        return index_1.Struct.initStructAt(0, ThirdPartyCapDescriptor, this);
    };
    CapDescriptor.prototype.isThirdPartyHosted = function () { return index_1.Struct.getUint16(0, this) === 5; };
    CapDescriptor.prototype.setThirdPartyHosted = function (value) {
        index_1.Struct.setUint16(0, 5, this);
        index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this));
    };
    CapDescriptor.prototype.toString = function () { return "CapDescriptor_" + _super.prototype.toString.call(this); };
    CapDescriptor.prototype.which = function () { return index_1.Struct.getUint16(0, this); };
    CapDescriptor.NONE = CapDescriptor_Which.NONE;
    CapDescriptor.SENDER_HOSTED = CapDescriptor_Which.SENDER_HOSTED;
    CapDescriptor.SENDER_PROMISE = CapDescriptor_Which.SENDER_PROMISE;
    CapDescriptor.RECEIVER_HOSTED = CapDescriptor_Which.RECEIVER_HOSTED;
    CapDescriptor.RECEIVER_ANSWER = CapDescriptor_Which.RECEIVER_ANSWER;
    CapDescriptor.THIRD_PARTY_HOSTED = CapDescriptor_Which.THIRD_PARTY_HOSTED;
    CapDescriptor._capnp = { displayName: "CapDescriptor", id: "8523ddc40b86b8b0", size: new index_1.ObjectSize(8, 1) };
    return CapDescriptor;
}(index_1.Struct));
exports.CapDescriptor = CapDescriptor;
var PromisedAnswer_Op_Which;
(function (PromisedAnswer_Op_Which) {
    PromisedAnswer_Op_Which[PromisedAnswer_Op_Which["NOOP"] = 0] = "NOOP";
    PromisedAnswer_Op_Which[PromisedAnswer_Op_Which["GET_POINTER_FIELD"] = 1] = "GET_POINTER_FIELD";
})(PromisedAnswer_Op_Which = exports.PromisedAnswer_Op_Which || (exports.PromisedAnswer_Op_Which = {}));
var PromisedAnswer_Op = /** @class */ (function (_super) {
    tslib_1.__extends(PromisedAnswer_Op, _super);
    function PromisedAnswer_Op() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PromisedAnswer_Op.prototype.isNoop = function () { return index_1.Struct.getUint16(0, this) === 0; };
    PromisedAnswer_Op.prototype.setNoop = function () { index_1.Struct.setUint16(0, 0, this); };
    PromisedAnswer_Op.prototype.getGetPointerField = function () {
        index_1.Struct.testWhich("getPointerField", index_1.Struct.getUint16(0, this), 1, this);
        return index_1.Struct.getUint16(2, this);
    };
    PromisedAnswer_Op.prototype.isGetPointerField = function () { return index_1.Struct.getUint16(0, this) === 1; };
    PromisedAnswer_Op.prototype.setGetPointerField = function (value) {
        index_1.Struct.setUint16(0, 1, this);
        index_1.Struct.setUint16(2, value, this);
    };
    PromisedAnswer_Op.prototype.toString = function () { return "PromisedAnswer_Op_" + _super.prototype.toString.call(this); };
    PromisedAnswer_Op.prototype.which = function () { return index_1.Struct.getUint16(0, this); };
    PromisedAnswer_Op.NOOP = PromisedAnswer_Op_Which.NOOP;
    PromisedAnswer_Op.GET_POINTER_FIELD = PromisedAnswer_Op_Which.GET_POINTER_FIELD;
    PromisedAnswer_Op._capnp = { displayName: "Op", id: "f316944415569081", size: new index_1.ObjectSize(8, 0) };
    return PromisedAnswer_Op;
}(index_1.Struct));
exports.PromisedAnswer_Op = PromisedAnswer_Op;
var PromisedAnswer = /** @class */ (function (_super) {
    tslib_1.__extends(PromisedAnswer, _super);
    function PromisedAnswer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PromisedAnswer.prototype.getQuestionId = function () { return index_1.Struct.getUint32(0, this); };
    PromisedAnswer.prototype.setQuestionId = function (value) { index_1.Struct.setUint32(0, value, this); };
    PromisedAnswer.prototype.adoptTransform = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    PromisedAnswer.prototype.disownTransform = function () { return index_1.Struct.disown(this.getTransform()); };
    PromisedAnswer.prototype.getTransform = function () { return index_1.Struct.getList(0, PromisedAnswer._Transform, this); };
    PromisedAnswer.prototype.hasTransform = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    PromisedAnswer.prototype.initTransform = function (length) { return index_1.Struct.initList(0, PromisedAnswer._Transform, length, this); };
    PromisedAnswer.prototype.setTransform = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    PromisedAnswer.prototype.toString = function () { return "PromisedAnswer_" + _super.prototype.toString.call(this); };
    PromisedAnswer.Op = PromisedAnswer_Op;
    PromisedAnswer._capnp = { displayName: "PromisedAnswer", id: "d800b1d6cd6f1ca0", size: new index_1.ObjectSize(8, 1) };
    return PromisedAnswer;
}(index_1.Struct));
exports.PromisedAnswer = PromisedAnswer;
var ThirdPartyCapDescriptor = /** @class */ (function (_super) {
    tslib_1.__extends(ThirdPartyCapDescriptor, _super);
    function ThirdPartyCapDescriptor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ThirdPartyCapDescriptor.prototype.adoptId = function (value) { index_1.Struct.adopt(value, index_1.Struct.getPointer(0, this)); };
    ThirdPartyCapDescriptor.prototype.disownId = function () { return index_1.Struct.disown(this.getId()); };
    ThirdPartyCapDescriptor.prototype.getId = function () { return index_1.Struct.getPointer(0, this); };
    ThirdPartyCapDescriptor.prototype.hasId = function () { return !index_1.Struct.isNull(index_1.Struct.getPointer(0, this)); };
    ThirdPartyCapDescriptor.prototype.setId = function (value) { index_1.Struct.copyFrom(value, index_1.Struct.getPointer(0, this)); };
    ThirdPartyCapDescriptor.prototype.getVineId = function () { return index_1.Struct.getUint32(0, this); };
    ThirdPartyCapDescriptor.prototype.setVineId = function (value) { index_1.Struct.setUint32(0, value, this); };
    ThirdPartyCapDescriptor.prototype.toString = function () { return "ThirdPartyCapDescriptor_" + _super.prototype.toString.call(this); };
    ThirdPartyCapDescriptor._capnp = { displayName: "ThirdPartyCapDescriptor", id: "d37007fde1f0027d", size: new index_1.ObjectSize(8, 1) };
    return ThirdPartyCapDescriptor;
}(index_1.Struct));
exports.ThirdPartyCapDescriptor = ThirdPartyCapDescriptor;
var Exception_Type;
(function (Exception_Type) {
    Exception_Type[Exception_Type["FAILED"] = 0] = "FAILED";
    Exception_Type[Exception_Type["OVERLOADED"] = 1] = "OVERLOADED";
    Exception_Type[Exception_Type["DISCONNECTED"] = 2] = "DISCONNECTED";
    Exception_Type[Exception_Type["UNIMPLEMENTED"] = 3] = "UNIMPLEMENTED";
})(Exception_Type = exports.Exception_Type || (exports.Exception_Type = {}));
var Exception = /** @class */ (function (_super) {
    tslib_1.__extends(Exception, _super);
    function Exception() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Exception.prototype.getReason = function () { return index_1.Struct.getText(0, this); };
    Exception.prototype.setReason = function (value) { index_1.Struct.setText(0, value, this); };
    Exception.prototype.getType = function () { return index_1.Struct.getUint16(4, this); };
    Exception.prototype.setType = function (value) { index_1.Struct.setUint16(4, value, this); };
    Exception.prototype.getObsoleteIsCallersFault = function () { return index_1.Struct.getBit(0, this); };
    Exception.prototype.setObsoleteIsCallersFault = function (value) { index_1.Struct.setBit(0, value, this); };
    Exception.prototype.getObsoleteDurability = function () { return index_1.Struct.getUint16(2, this); };
    Exception.prototype.setObsoleteDurability = function (value) { index_1.Struct.setUint16(2, value, this); };
    Exception.prototype.toString = function () { return "Exception_" + _super.prototype.toString.call(this); };
    Exception.Type = Exception_Type;
    Exception._capnp = { displayName: "Exception", id: "d625b7063acf691a", size: new index_1.ObjectSize(8, 1) };
    return Exception;
}(index_1.Struct));
exports.Exception = Exception;
Payload._CapTable = capnp.CompositeList(CapDescriptor);
PromisedAnswer._Transform = capnp.CompositeList(PromisedAnswer_Op);
