// capnp-ts/lib-test/integration/serialization-demo.capnp.js
"use strict";
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * This file has been automatically generated by the [capnpc-ts utility](https://github.com/jdiaz5513/capnp-ts).
 */
var capnp = require("capnp-ts");
var capnp_ts_1 = require("capnp-ts");
exports._capnpFileId = "b597bf4897e54f89";
var Person_PhoneNumber_Type;
(function (Person_PhoneNumber_Type) {
    Person_PhoneNumber_Type[Person_PhoneNumber_Type["MOBILE"] = 0] = "MOBILE";
    Person_PhoneNumber_Type[Person_PhoneNumber_Type["HOME"] = 1] = "HOME";
    Person_PhoneNumber_Type[Person_PhoneNumber_Type["WORK"] = 2] = "WORK";
})(Person_PhoneNumber_Type = exports.Person_PhoneNumber_Type || (exports.Person_PhoneNumber_Type = {}));
var Person_PhoneNumber = /** @class */ (function (_super) {
    tslib_1.__extends(Person_PhoneNumber, _super);
    function Person_PhoneNumber() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Person_PhoneNumber.prototype.getNumber = function () { return capnp_ts_1.Struct.getText(0, this); };
    Person_PhoneNumber.prototype.setNumber = function (value) { capnp_ts_1.Struct.setText(0, value, this); };
    Person_PhoneNumber.prototype.getType = function () { return capnp_ts_1.Struct.getUint16(0, this); };
    Person_PhoneNumber.prototype.setType = function (value) { capnp_ts_1.Struct.setUint16(0, value, this); };
    Person_PhoneNumber.prototype.toString = function () { return "Person_PhoneNumber_" + _super.prototype.toString.call(this); };
    Person_PhoneNumber.Type = Person_PhoneNumber_Type;
    Person_PhoneNumber._capnp = { displayName: "PhoneNumber", id: "cba8ed6b45001ccc", size: new capnp_ts_1.ObjectSize(8, 1) };
    return Person_PhoneNumber;
}(capnp_ts_1.Struct));
exports.Person_PhoneNumber = Person_PhoneNumber;
var Person_Employment_Which;
(function (Person_Employment_Which) {
    Person_Employment_Which[Person_Employment_Which["UNEMPLOYED"] = 0] = "UNEMPLOYED";
    Person_Employment_Which[Person_Employment_Which["EMPLOYER"] = 1] = "EMPLOYER";
    Person_Employment_Which[Person_Employment_Which["SCHOOL"] = 2] = "SCHOOL";
    Person_Employment_Which[Person_Employment_Which["SELF_EMPLOYED"] = 3] = "SELF_EMPLOYED";
})(Person_Employment_Which = exports.Person_Employment_Which || (exports.Person_Employment_Which = {}));
var Person_Employment = /** @class */ (function (_super) {
    tslib_1.__extends(Person_Employment, _super);
    function Person_Employment() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Person_Employment.prototype.isUnemployed = function () { return capnp_ts_1.Struct.getUint16(4, this) === 0; };
    Person_Employment.prototype.setUnemployed = function () { capnp_ts_1.Struct.setUint16(4, 0, this); };
    Person_Employment.prototype.getEmployer = function () {
        capnp_ts_1.Struct.testWhich("employer", capnp_ts_1.Struct.getUint16(4, this), 1, this);
        return capnp_ts_1.Struct.getText(3, this);
    };
    Person_Employment.prototype.isEmployer = function () { return capnp_ts_1.Struct.getUint16(4, this) === 1; };
    Person_Employment.prototype.setEmployer = function (value) {
        capnp_ts_1.Struct.setUint16(4, 1, this);
        capnp_ts_1.Struct.setText(3, value, this);
    };
    Person_Employment.prototype.getSchool = function () {
        capnp_ts_1.Struct.testWhich("school", capnp_ts_1.Struct.getUint16(4, this), 2, this);
        return capnp_ts_1.Struct.getText(3, this);
    };
    Person_Employment.prototype.isSchool = function () { return capnp_ts_1.Struct.getUint16(4, this) === 2; };
    Person_Employment.prototype.setSchool = function (value) {
        capnp_ts_1.Struct.setUint16(4, 2, this);
        capnp_ts_1.Struct.setText(3, value, this);
    };
    Person_Employment.prototype.isSelfEmployed = function () { return capnp_ts_1.Struct.getUint16(4, this) === 3; };
    Person_Employment.prototype.setSelfEmployed = function () { capnp_ts_1.Struct.setUint16(4, 3, this); };
    Person_Employment.prototype.toString = function () { return "Person_Employment_" + _super.prototype.toString.call(this); };
    Person_Employment.prototype.which = function () { return capnp_ts_1.Struct.getUint16(4, this); };
    Person_Employment.UNEMPLOYED = Person_Employment_Which.UNEMPLOYED;
    Person_Employment.EMPLOYER = Person_Employment_Which.EMPLOYER;
    Person_Employment.SCHOOL = Person_Employment_Which.SCHOOL;
    Person_Employment.SELF_EMPLOYED = Person_Employment_Which.SELF_EMPLOYED;
    Person_Employment._capnp = { displayName: "employment", id: "927f49708287c3b6", size: new capnp_ts_1.ObjectSize(8, 4) };
    return Person_Employment;
}(capnp_ts_1.Struct));
exports.Person_Employment = Person_Employment;
var Person = /** @class */ (function (_super) {
    tslib_1.__extends(Person, _super);
    function Person() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Person.prototype.getId = function () { return capnp_ts_1.Struct.getUint32(0, this); };
    Person.prototype.setId = function (value) { capnp_ts_1.Struct.setUint32(0, value, this); };
    Person.prototype.getName = function () { return capnp_ts_1.Struct.getText(0, this); };
    Person.prototype.setName = function (value) { capnp_ts_1.Struct.setText(0, value, this); };
    Person.prototype.getEmail = function () { return capnp_ts_1.Struct.getText(1, this); };
    Person.prototype.setEmail = function (value) { capnp_ts_1.Struct.setText(1, value, this); };
    Person.prototype.adoptPhones = function (value) { capnp_ts_1.Struct.adopt(value, capnp_ts_1.Struct.getPointer(2, this)); };
    Person.prototype.disownPhones = function () { return capnp_ts_1.Struct.disown(this.getPhones()); };
    Person.prototype.getPhones = function () { return capnp_ts_1.Struct.getList(2, Person._Phones, this); };
    Person.prototype.hasPhones = function () { return !capnp_ts_1.Struct.isNull(capnp_ts_1.Struct.getPointer(2, this)); };
    Person.prototype.initPhones = function (length) { return capnp_ts_1.Struct.initList(2, Person._Phones, length, this); };
    Person.prototype.setPhones = function (value) { capnp_ts_1.Struct.copyFrom(value, capnp_ts_1.Struct.getPointer(2, this)); };
    Person.prototype.getEmployment = function () { return capnp_ts_1.Struct.getAs(Person_Employment, this); };
    Person.prototype.initEmployment = function () { return capnp_ts_1.Struct.getAs(Person_Employment, this); };
    Person.prototype.toString = function () { return "Person_" + _super.prototype.toString.call(this); };
    Person.PhoneNumber = Person_PhoneNumber;
    Person._capnp = { displayName: "Person", id: "efbbc4e996f07104", size: new capnp_ts_1.ObjectSize(8, 4) };
    return Person;
}(capnp_ts_1.Struct));
exports.Person = Person;
var AddressBook = /** @class */ (function (_super) {
    tslib_1.__extends(AddressBook, _super);
    function AddressBook() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AddressBook.prototype.adoptPeople = function (value) { capnp_ts_1.Struct.adopt(value, capnp_ts_1.Struct.getPointer(0, this)); };
    AddressBook.prototype.disownPeople = function () { return capnp_ts_1.Struct.disown(this.getPeople()); };
    AddressBook.prototype.getPeople = function () { return capnp_ts_1.Struct.getList(0, AddressBook._People, this); };
    AddressBook.prototype.hasPeople = function () { return !capnp_ts_1.Struct.isNull(capnp_ts_1.Struct.getPointer(0, this)); };
    AddressBook.prototype.initPeople = function (length) { return capnp_ts_1.Struct.initList(0, AddressBook._People, length, this); };
    AddressBook.prototype.setPeople = function (value) { capnp_ts_1.Struct.copyFrom(value, capnp_ts_1.Struct.getPointer(0, this)); };
    AddressBook.prototype.toString = function () { return "AddressBook_" + _super.prototype.toString.call(this); };
    AddressBook._capnp = { displayName: "AddressBook", id: "f724540a01e98224", size: new capnp_ts_1.ObjectSize(0, 1) };
    return AddressBook;
}(capnp_ts_1.Struct));
exports.AddressBook = AddressBook;
Person._Phones = capnp.CompositeList(Person_PhoneNumber);
AddressBook._People = capnp.CompositeList(Person);
