// capnp-ts/lib-test/integration/test-import.capnp.js
"use strict";
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var capnp_ts_1 = require("capnp-ts");
var test_capnp_1 = require("./test.capnp");
exports._capnpFileId = "f36d7b330303c66e";
var TestImport = /** @class */ (function (_super) {
    tslib_1.__extends(TestImport, _super);
    function TestImport() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestImport.prototype.adoptField = function (value) { capnp_ts_1.Struct.adopt(value, capnp_ts_1.Struct.getPointer(0, this)); };
    TestImport.prototype.disownField = function () { return capnp_ts_1.Struct.disown(this.getField()); };
    TestImport.prototype.getField = function () { return capnp_ts_1.Struct.getStruct(0, test_capnp_1.TestAllTypes, this); };
    TestImport.prototype.hasField = function () { return !capnp_ts_1.Struct.isNull(capnp_ts_1.Struct.getPointer(0, this)); };
    TestImport.prototype.initField = function () { return capnp_ts_1.Struct.initStructAt(0, test_capnp_1.TestAllTypes, this); };
    TestImport.prototype.setField = function (value) { capnp_ts_1.Struct.copyFrom(value, capnp_ts_1.Struct.getPointer(0, this)); };
    TestImport.prototype.toString = function () { return "TestImport_" + _super.prototype.toString.call(this); };
    TestImport._capnp = { displayName: "TestImport", id: "bc55b08b672b5d97", size: new capnp_ts_1.ObjectSize(0, 1) };
    return TestImport;
}(capnp_ts_1.Struct));
exports.TestImport = TestImport;
